src/main/java/com/google/api/generator/engine/ast/CastExpr.java

```diff
@@ -52,17 +52,16 @@ public CastExpr build() {
       // Only numeric types can be casted around.
       if (castType.isPrimitiveType() && exprType.isPrimitiveType()) {
         // Check that the types are either both booleans or both not booleans.
-        // TODO(xiaozhenliu): This is one of the places where we need to check the equality of boxed
-        // and primitive types.
         Preconditions.checkState(
             !(castType.equals(TypeNode.BOOLEAN) ^ exprType.equals(TypeNode.BOOLEAN)),
             "Numeric and boolean types are not inter-castable");
       } else {
         Preconditions.checkState(
-            TypeNode.isReferenceType(castExpr.type())
-                && (castExpr.expr().type().equals(TypeNode.NULL)
-                    || TypeNode.isReferenceType(castExpr.expr().type())),
-            "Reference types can only be casted to reference types");
+            castExpr.type().equals(castExpr.expr().type())
```

Comment:
Could we add some parentheses to clarify the order? This currently reads `( cond1 || cond2 && (cond3))`, so the intent of the first three aren't clear.

Reply:
Since this is a long condition, could we please put sub-condition into named variables to the overall one is easier to read?

Reply:
Sounds cool, defined a `isReferenceTypeEquals` to check the condition when castType is a reference type.